rust/arrow/src/json/reader.rs

```diff
@@ -0,0 +1,853 @@
+                        }
+                        Value::String(_) => {
+                            if values.contains_key(k) {
+                                let x = values.get_mut(k).unwrap();
+                                x.insert(DataType::Utf8);
+                            } else {
+                                // create hashset and add value type
+                                let mut hs = HashSet::new();
+                                hs.insert(DataType::Utf8);
+                                values.insert(k.to_string(), hs);
+                            }
+                        }
+                        Value::Object(_) => {
+                            panic!("Reading nested JSON structes currently not supported")
+                        }
+                    }
+                });
+            }
+            _ => {
+                // return an error, we expect a value
+            }
+        };
+    }
+
+    let schema = generate_schema(values);
+
+    // return the reader seek back to the start
+    &reader.into_inner().seek(SeekFrom::Start(0))?;
+
+    Ok(schema)
+}
+
+/// JSON file reader
+pub struct Reader<R: Read> {
+    /// Explicit schema for the JSON file
+    schema: Arc<Schema>,
+    /// Optional projection for which columns to load (case-sensitive names)
+    projection: Option<Vec<String>>,
+    /// File reader
+    reader: BufReader<R>,
+    /// Batch size (number of records to load each time)
+    batch_size: usize,
+}
+
+impl<R: Read> Reader<R> {
+    /// Create a new JSON Reader from any value that implements the `Read` trait.
+    ///
+    /// If reading a `File`, you can customise the Reader, such as to enable schema
+    /// inference, use `ReaderBuilder`.
+    pub fn new(
+        reader: BufReader<R>,
+        schema: Arc<Schema>,
+        batch_size: usize,
+        projection: Option<Vec<String>>,
+    ) -> Self {
+        Self {
+            schema,
+            projection,
+            reader,
+            batch_size,
+        }
+    }
+
+    /// Read the next batch of records
+    pub fn next(&mut self) -> Result<Option<RecordBatch>> {
+        let mut rows: Vec<Value> = Vec::with_capacity(self.batch_size);
+        let mut line = String::new();
+        for _ in 0..self.batch_size {
+            self.reader.read_line(&mut line)?;
+            if !line.is_empty() {
+                rows.push(serde_json::from_str(&line).expect("Not valid JSON"));
+                line = String::new();
+            } else {
+                break;
+            }
+        }
+
+        let rows = &rows[..];
+        let projection = self.projection.clone().unwrap_or(vec![]);
+        let arrays: Result<Vec<ArrayRef>> = self
+            .schema
+            .clone()
+            .fields()
+            .iter()
+            .filter(|field| {
+                if projection.is_empty() {
+                    return true;
+                }
+                projection.contains(field.name())
```

Comment:
This seems to picking the fields that are in the projection correctly but not necessarily in the correct order.

Reply:
I need to enable preserving field order on serde_json to keep field order. I opened a separate JIRA for this yesterday. I'd like to address it separately because it also impacts data type JSON conversions.

Reply:
That's fine. I think for now we can just ensure that we pass the projection in the correct order (I'm fairly sure that is already the case).